src/font_renderer/ct_font_renderer.cpp

```diff
@@ -0,0 +1,144 @@
+
+#if DPFB_USE_CORETEXT
+
+#include "font_renderer/font_renderer.h"
+
+#include <CoreFoundation/CoreFoundation.h>
+#include <CoreGraphics/CoreGraphics.h>
+#include <CoreText/CoreText.h>
+
+
+
+class CtFontRenderer : public FontRenderer {
+public:
+    explicit CtFontRenderer(const FontRendererArgs& args);
+
+    FontMetrics getFontMetrics() const override;
+
+    GlyphIndex getGlyphIndex(char32_t cp) const override;
+    GlyphMetrics getGlyphMetrics(GlyphIndex glyphIdx) const override;
+    void renderGlyph(GlyphIndex glyphIdx, Image& image) const override;
+    virtual ~CtFontRenderer() override;
+private:
+    CTFontRef font;
+};
+
+
+CtFontRenderer::CtFontRenderer(const FontRendererArgs& args)
+{
+    CTFontDescriptorRef descriptor = CTFontManagerCreateFontDescriptorFromData(CFDataCreate(nullptr, static_cast<const UInt8*>(args.data), args.dataSize));
+    if (!descriptor)
+        throw FontRendererError("CoreText can't create descriptor");
+    font = CTFontCreateWithFontDescriptor(descriptor, static_cast<CGFloat>(args.pxSize), nullptr);
+    CFRelease(descriptor);
+    if (!font)
+        throw FontRendererError("CoreText can't init font");
+}
+
+
+FontMetrics CtFontRenderer::getFontMetrics() const
+{
+    FontMetrics metrics;
+
+    metrics.ascender = ceil(CTFontGetAscent(font));
+    metrics.descender = ceil(CTFontGetDescent(font));
+    metrics.lineHeight = ceil(CTFontGetLeading(font) + CTFontGetAscent(font) + CTFontGetDescent(font));
+
+    return metrics;
+}
+
+
+GlyphIndex CtFontRenderer::getGlyphIndex(char32_t cp) const
+{
+    CGGlyph glyph;
+    UniChar ch = static_cast<UniChar>(cp);
+    if (CTFontGetGlyphsForCharacters(font, &ch, &glyph, 1))
+        return static_cast<GlyphIndex>(glyph);
+    return 0;
+}
+
+#define CORETEXT_SHIFT_X_POSITION 1
```

Comment:
It's better to leave only the used parts of code and remove the macro.

Reply:
Removed with c2a3996